tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/Inliner.scala

```diff
@@ -0,0 +1,238 @@
+      // 1) We never want to remove the definitions from a call-by-name or lambda, so the operator filter depends on
+      // whether or not this is a generic LET-IN or a call-by-name
+      // 2) As `transform` also embeds call-by-names, we don't want to embed an already embedded call-by-name
+      val (filterFun, bodyTx): (DeclFilter, (Scope, TlaEx) => TlaEx) =
+        if (isCallByName(letInEx))
+          (FilterFun.ALL, { case (_, e) => e }) // (ALL, NO-OP)
+        else
+          (negateFilter(nonNullaryFilter), { case (s, e) => transform(s)(e) }) // (!NULLARY, transform)
+
+      val (newScope, remainingOpers) = pushDeclsIntoScope(filterFun)(scope, defs)
+      // pushDeclsIntoScope is generic, so it doesn't know all TlaDecl are TlaOperDecl, so we just cast all of them
+      val castDecls = remainingOpers.map(_.asInstanceOf[TlaOperDecl])
+      // if we know this is already a call-by name, no need to re-embed, since body must be a NameEx
+      // In this case newBody = body
+      val newBody = bodyTx(newScope, body)
+
+      // If we ended up pushing all declarations into the scope, then we can get rid of toplevel LET-IN
+      if (castDecls.isEmpty) newBody
+      else LetInEx(newBody, castDecls: _*)(newBody.typeTag)
+
+    // call-by-name
+    case nameEx: NameEx if scope.contains(nameEx.name) =>
+      embedCallByName(scope)(nameEx)
+
+    // Standard recursive case
+    case ex @ OperEx(oper, args @ _*) =>
+      OperEx(oper, args.map(transform(scope)): _*)(ex.typeTag)
+
+    // No inlining for non-operator non-name expressions
+    case ex => ex
+  }
+
+  // Module-level transformation. Guarantees traversal of declarations in dependency order
+  override def apply(m: TlaModule): TlaModule = {
+    val declSorter = new DeclarationSorter
+    // Instead of pre-building a body map, sorts operators according to
+    // dependency order and iterates over a sorted collection
+    val moduleWithSortedDecls = declSorter(m)
+
+    val sortedDecls = moduleWithSortedDecls.declarations
+
+    // Instead of building a parallel BodyMap, we build scope iteratively, as the declarations are now in order.
+    // Because we keep global declarations, we may need to use FilterFun.ALL or a similar filter unrelated to arity
+    // Also, because global operators are inlined even if nullary, we force scope filter to be ALL
+    val inlinedDecls = pushDeclsIntoScope(toplevelFilter, FilterFun.ALL)(emptyScope, sortedDecls)._2
+
+    m.copy(declarations = inlinedDecls)
+  }
+
+}
+
+object Inliner {
+
+  type Scope = BodyMap
+  def emptyScope: Scope = Map.empty
+
+  type DeclFilter = TlaOperDecl => Boolean
+
+  def negateFilter(df: DeclFilter): DeclFilter = d => !df(d)
+
+  object FilterFun {
+    def ALL: DeclFilter = _ => true
+    def NAMED(set: Set[String]): DeclFilter = d => set.contains(d.name)
+  }
+
+  // Checks for the specific LET-IN form used by LAMBDA and call-by-name
+  def isCallByName(letInEx: LetInEx): Boolean = letInEx match {
```

Comment:
As above, is this here because it's static?
```suggestion
  private def isCallByName(letInEx: LetInEx): Boolean = letInEx match {
```

Reply:
Why? Also, Not sure if this will even work with `import Inliner._` in the class above.

Reply:
It'd work, companion object/class have access to each other's private fields.
But I just saw that this is used in the unit tests, never mind.